runtime/server/batch_query.go

```diff
@@ -0,0 +1,251 @@
+package server
+
+import (
+	"context"
+	"sync"
+
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+)
+
+func (s *Server) QueryBatch(req *runtimev1.QueryBatchRequest, srv runtimev1.QueryService_QueryBatchServer) error {
+	ctx, cancel := context.WithCancelCause(srv.Context())
+	cancelled := false
+
+	// TODO: Performance improvements:
+	//       1. Check for access upfront based on what is in the request
+	//       2. Check for cache and return those queries immediately before creating a goroutine
+	//       3. Use a goroutine pool with size equal to driver's concurrency to execute the queries
+
+	var wg sync.WaitGroup
+
+	for _, query := range req.Queries {
+		wg.Add(1)
+		go func(queryEntry *runtimev1.QueryBatchEntry) {
+			defer wg.Done()
+
+			resp := s.forwardQuery(ctx, req, queryEntry)
+
+			if err := srv.Send(resp); err != nil && !cancelled {
+				// if we failed to send response, cancel the context so any pending queries are cancelled
+				cancel(err)
+				cancelled = true
+			}
+		}(query)
+	}
+
+	wg.Wait()
+	return nil
+}
+
+func (s *Server) forwardQuery(ctx context.Context, query *runtimev1.QueryBatchRequest, queryEntry *runtimev1.QueryBatchEntry) *runtimev1.QueryBatchResponse {
+	resp := &runtimev1.QueryBatchResponse{
+		Key: queryEntry.Key,
+	}
+
+	var err error
+	switch typedQueryEntry := queryEntry.Query.(type) {
+	case *runtimev1.QueryBatchEntry_MetricsViewToplistRequest:
+		var r *runtimev1.MetricsViewToplistResponse
+		if typedQueryEntry.MetricsViewToplistRequest.Priority == 0 {
+			typedQueryEntry.MetricsViewToplistRequest.Priority = query.Priority
+		}
```

Comment:
See earlier comment about moving priority to the entry. If the frontend doesn't need the conditional, I think it's okay to just always use the value from the entry

Reply:
Ya i will remove these. Frontend can send the same priority to all requests if needed. Let that control be on the sender.